src/sinks/util/service.rs

```diff
@@ -269,6 +291,54 @@ impl TowerRequestSettings {
             .service(service);
         BatchSink::new(service, batch, batch_timeout)
     }
+
+    pub fn distributed_service<Req, RL, S, H, K, D>(
+        self,
+        retry_logic: RL,
+        services: D,
+        reactivate_delay: Duration,
+    ) -> DistributedService<S, RL, K, Req>
+    where
+        Req: Clone + Send + 'static,
+        RL: RetryLogic<Response = S::Response>,
+        S: Service<Req> + Clone + Send + 'static,
+        S::Error: Into<crate::Error> + Send + Sync + 'static,
+        S::Response: Send,
+        S::Future: Send + 'static,
+        H: Fn() -> BoxFuture<'static, bool> + Send + 'static,
+        K: Eq + Hash + Clone + Copy + Send + 'static,
+        D: TryStream<Ok = Change<K, (S, H)>, Error = crate::Error> + Send + 'static,
+    {
+        let service_layer = ServiceBuilder::new().timeout(self.timeout);
+
+        let logic = retry_logic.clone();
+        let open = OpenGauge::new();
+        let services = services.map_ok(move |change| match change {
+            Change::Insert(key, (service, healthcheck)) => Change::Insert(
+                key,
+                HealthService::new(
+                    service_layer.service(service),
+                    healthcheck,
+                    logic.clone(),
+                    reactivate_delay,
+                    open.clone(),
+                ),
+            ),
+            Change::Remove(key) => Change::Remove(key),
+        });
+
+        ServiceBuilder::new()
+            .layer(TowerRequestLayer {
+                settings: self,
+                retry_logic,
+                _pd: std::marker::PhantomData,
+            })
+            // There are other mechanisms limiting the number of concurrent requests/poll_ready
+            // so we don't need to do that here, hence usize::MAX. But underlying library supports less
+            // so >>3 is used as per tokio semaphore documentation.
+            .layer(BufferLayer::new(usize::MAX >> 3))
```

Comment:
The passed bound isn't bounding messages but requests, and as commented other mechanisms/layers handle that by applying backpressure which will reach BufferLayer and stop it to queue further requests. In this case ARC is that layer. 

Also while it's possible to compute upper bound on concurrency at the moment that won't be the case once snooping is added.

Reply:
No, it's bounding messages going into the buffer layer: https://docs.rs/tower/latest/src/tower/buffer/service.rs.html#91

They're using an unbounded channel with a semaphore since `Semaphore` has a `poll_*`-style API available. If you look through `poll_ready`, it never polls the underlying service on the other side, so all it's doing is limiting callers if the channel is full, and since the channel size is only limited by `bound`...

Reply:
... good call. I'll add it.